src/glang/codegen/builtin_callables.py

```diff
@@ -606,6 +609,111 @@ def assert_can_write_to(self) -> None:
         self._src_expr.assert_can_write_to()
 
 
+class ArrayIndexExpression(BuiltinCallable):
+    def __init__(
+        self,
+        specialization: list[SpecializationItem],
+        arguments: list[TypedExpression],
+        meta: Meta,
+    ) -> None:
+        assert len(specialization) == 0
+        assert len(arguments) >= 2
+
+        array_ptr, *indices = arguments
+
+        assert array_ptr.has_address
+
+        self._type_of_array: Type = array_ptr.result_type.convert_to_value_type()
+        self._array_ptr = array_ptr
+
+        array_definition = self._type_of_array.definition
+        assert isinstance(array_definition, StackArrayDefinition | HeapArrayDefinition)
+
+        if isinstance(array_definition, StackArrayDefinition):
+            assert len(array_definition.dimensions) == len(indices)
+
+        else:
+            assert len(indices) == 1 + len(array_definition.known_dimensions)
+
+        self._element_type: Type = array_definition.member
+        self._conversion_exprs: list[TypedExpression] = []
+
+        # Now convert all the indices into integers using standard implicit rules
+        self._indices: list[TypedExpression] = []
+        for index in indices:
+            index_expr, conversions = do_implicit_conversion(
+                index, SizeType(), "array index access"
+            )
+            self._indices.append(index_expr)
+            self._conversion_exprs.extend(conversions)
+
+        result_type = array_ptr.result_type_as_if_borrowed
+        StaticTypedExpression.__init__(
```

Comment:
This feels needlessly error prone; can we not use `super()` instead? (I guess we'll need to inherit from `ABC` instead of making the `BuiltinCallable` constructor abstract, hopefully that works)